File: viewer-macos/Sources/AppControlViewer/Security/Handshake.swift
import Crypto
import Foundation

enum HandshakeRole { case host, viewer }

struct HandshakeMessage: Codable {
    var t: String = "hs"
    var v: String = "ac/1"
    var ephemeral: String       // base64, 32 Byte X25519
    var `static`: String        // base64, 32 Byte X25519 (Langzeit-Identität)
    var name: String
    var nonce: String           // base64, 16 Byte

    enum CodingKeys: String, CodingKey {
        case t, v, ephemeral, name, nonce
        case `static` = "static"
    }
}

struct HandshakeResult {
    let sessionKey: Data
    let peerStaticPublicKey: Data
    let peerName: String
    let sasEmoji: String
    let sasIndices: [Int]
}

/// Triple-Diffie-Hellman-Handshake, Struktur von Noise `KK` mit vorgeschaltetem
/// PSK-Umschlag.
///
/// Die drei DH-Operationen leisten zusammen:
/// - `dh_1` (e × e): Forward Secrecy — aufgezeichnete Sitzungen bleiben unlesbar,
///   selbst wenn später beide statischen Schlüssel kompromittiert werden
/// - `dh_2` (e × s): bindet die Sitzung an die statische Identität des Viewers
/// - `dh_3` (s × e): bindet sie an die des Hosts — beidseitige Authentifizierung
///
/// Gegenstück: `Handshake.cs`. Siehe docs/05-security.md §5.3.
enum Handshake {

    /// Phase-A-Schlüssel: leitet sich allein aus dem Pairing-Ticket ab.
    static func deriveHandshakeKey(psk: Data, roomId: Data) -> Data {
        hkdf(ikm: psk, salt: roomId, info: "ac/1 handshake", length: 32)
    }

    /// Berechnet den Sitzungsschlüssel aus den eigenen privaten und den
    /// empfangenen öffentlichen Schlüsseln.
    static func computeSession(
        role: HandshakeRole,
        ownEphemeral: Curve25519.KeyAgreement.PrivateKey,
        ownStatic: Curve25519.KeyAgreement.PrivateKey,
        peerEphemeralPublic: Data,
        peerStaticPublic: Data,
        nonceHost: Data,
        nonceViewer: Data,
        peerName: String
    ) throws -> HandshakeResult {

        let peerEph = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerEphemeralPublic)
        let peerStatic = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerStaticPublic)

        let dh1 = try ownEphemeral.sharedSecretFromKeyAgreement(with: peerEph)

        // Die Rollenabhängigkeit ist der Punkt, an dem eine Implementierung am
        // ehesten von der anderen abweicht — deshalb gibt es dafür Testvektoren.
        let dh2: SharedSecret
        let dh3: SharedSecret
        switch role {
        case .host:
            dh2 = try ownEphemeral.sharedSecretFromKeyAgreement(with: peerStatic)  // e_H × s_V
            dh3 = try ownStatic.sharedSecretFromKeyAgreement(with: peerEph)        // s_H × e_V
        case .viewer:
            dh2 = try ownStatic.sharedSecretFromKeyAgreement(with: peerEph)        // s_V × e_H
            dh3 = try ownEphemeral.sharedSecretFromKeyAgreement(with: peerStatic)  // e_V × s_H
        }

        var ikm = Data()
        dh1.withUnsafeBytes { ikm.append(contentsOf: $0) }
        dh2.withUnsafeBytes { ikm.append(contentsOf: $0) }
        dh3.withUnsafeBytes { ikm.append(contentsOf: $0) }

        // Salt-Reihenfolge ist IMMER Host zuerst, unabhängig davon, wer sendet.
        var saltInput = Data()
        saltInput.append(nonceHost)
        saltInput.append(nonceViewer)
        let salt = Data(SHA256.hash(data: saltInput))

        let sessionKey = hkdf(ikm: ikm, salt: salt, info: "ac/1 session", length: 32)
        let (emoji, indices) = computeSas(sessionKey: sessionKey)

        return HandshakeResult(
            sessionKey: sessionKey,
            peerStaticPublicKey: peerStaticPublic,
            peerName: peerName,
            sasEmoji: emoji,
            sasIndices: indices)
    }

    /// Fünf Emojis aus einer 64er-Tabelle = 30 Bit.
    ///
    /// Beide Seiten zeigen dieselben Symbole; die Nutzer vergleichen sie über
    /// einen **anderen** Kanal. Ein MitM müsste zwei verschiedene
    /// Sitzungsschlüssel etablieren, die zufällig dieselben fünf Emojis ergeben:
    /// 1 zu 2³⁰. Siehe docs/05-security.md §5.4.
    static func computeSas(sessionKey: Data) -> (emoji: String, indices: [Int]) {
        let raw = hkdf(ikm: sessionKey, salt: Data(), info: "ac/1 sas", length: 5)
        let indices = raw.map { Int($0) % 64 }
        let emoji = indices.map { EmojiTable.symbols[$0] }.joined(separator: " ")
        return (emoji, indices)
    }

    /// Anzeigefreundlicher Fingerprint eines statischen Schlüssels.
    static func fingerprint(_ staticPublicKey: Data) -> String {
        let hash = Array(SHA256.hash(data: staticPublicKey).prefix(10))
        return stride(from: 0, to: 10, by: 2)
            .map { String(format: "%02x%02x", hash[$0], hash[$0 + 1]) }
            .joined(separator: " ")
    }

    static func hkdf(ikm: Data, salt: Data, info: String, length: Int) -> Data {
        let derived = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: ikm),
            salt: salt,
            info: Data(info.utf8),
            outputByteCount: length)
        return derived.withUnsafeBytes { Data($0) }
    }
}

File: host-windows/src/AppControl.Host/Security/Handshake.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using NSec.Cryptography;

namespace AppControl.Host.Security;

/// <summary>Rolle im Handshake. Bestimmt die Zuordnung der DH-Operationen.</summary>
public enum HandshakeRole { Host, Viewer }

public sealed record HandshakeMessage
{
    [JsonPropertyName("t")]         public string Type { get; init; } = "hs";
    [JsonPropertyName("v")]         public string Version { get; init; } = "ac/1";
    [JsonPropertyName("ephemeral")] public string EphemeralPublicKey { get; init; } = "";
    [JsonPropertyName("static")]    public string StaticPublicKey { get; init; } = "";
    [JsonPropertyName("name")]      public string Name { get; init; } = "";
    [JsonPropertyName("nonce")]     public string Nonce { get; init; } = "";
}

public sealed record HandshakeResult(
    byte[] SessionKey,
    byte[] PeerStaticPublicKey,
    string PeerName,
    string SasEmoji,
    int[] SasIndices);

/// <summary>
/// Triple-Diffie-Hellman-Handshake, Struktur von Noise KK mit vorgeschaltetem
/// PSK-Umschlag.
///
/// Die drei DH-Operationen leisten zusammen:
///   dh_1 (e x e)  Forward Secrecy - aufgezeichnete Sitzungen bleiben unlesbar,
///                 selbst wenn spaeter beide statischen Schluessel kompromittiert
///                 werden
///   dh_2 (e x s)  bindet die Sitzung an die statische Identitaet des Viewers
///   dh_3 (s x e)  bindet sie an die des Hosts - beidseitige Authentifizierung
///
/// Siehe docs/05-security.md §5.3. Gegenstueck: Security/Handshake.swift.
/// </summary>
public static class Handshake
{
    private static readonly KeyAgreementAlgorithm X25519 = KeyAgreementAlgorithm.X25519;

    /// <summary>Phase-A-Schluessel: leitet sich allein aus dem Pairing-Ticket ab.</summary>
    public static byte[] DeriveHandshakeKey(ReadOnlySpan<byte> psk, ReadOnlySpan<byte> roomId)
        => Hkdf(psk, roomId, "ac/1 handshake", 32);

    /// <summary>
    /// Berechnet den Sitzungsschluessel aus den eigenen privaten und den
    /// empfangenen oeffentlichen Schluesseln.
    /// </summary>
    public static HandshakeResult ComputeSession(
        HandshakeRole role,
        Key ownEphemeral,
        Key ownStatic,
        ReadOnlySpan<byte> peerEphemeralPublic,
        ReadOnlySpan<byte> peerStaticPublic,
        ReadOnlySpan<byte> nonceHost,
        ReadOnlySpan<byte> nonceViewer,
        string peerName)
    {
        var peerEph = PublicKey.Import(X25519, peerEphemeralPublic, KeyBlobFormat.RawPublicKey);
        var peerStatic = PublicKey.Import(X25519, peerStaticPublic, KeyBlobFormat.RawPublicKey);

        var dh1 = Agree(ownEphemeral, peerEph);

        // Die Rollenabhaengigkeit ist der Punkt, an dem eine Implementierung am
        // ehesten von der anderen abweicht - deshalb gibt es dafuer Testvektoren.
        byte[] dh2, dh3;
        if (role == HandshakeRole.Host)
        {
            dh2 = Agree(ownEphemeral, peerStatic);   // e_H x s_V
            dh3 = Agree(ownStatic, peerEph);         // s_H x e_V
        }
        else
        {
            dh2 = Agree(ownStatic, peerEph);         // s_V x e_H  ==  e_H x s_V
            dh3 = Agree(ownEphemeral, peerStatic);   // e_V x s_H  ==  s_H x e_V
        }

        var ikm = new byte[96];
        dh1.CopyTo(ikm, 0);
        dh2.CopyTo(ikm, 32);
        dh3.CopyTo(ikm, 64);

        // Salt-Reihenfolge ist IMMER Host zuerst, unabhaengig davon, wer sendet.
        var salt = SHA256.HashData([.. nonceHost, .. nonceViewer]);

        var sessionKey = Hkdf(ikm, salt, "ac/1 session", 32);

        CryptographicOperations.ZeroMemory(dh1);
        CryptographicOperations.ZeroMemory(dh2);
        CryptographicOperations.ZeroMemory(dh3);
        CryptographicOperations.ZeroMemory(ikm);

        var (emoji, indices) = ComputeSas(sessionKey);

        return new HandshakeResult(
            sessionKey, peerStaticPublic.ToArray(), peerName, emoji, indices);
    }

    private static byte[] Agree(Key ownPrivate, PublicKey peerPublic)
    {
        var shared = X25519.Agree(ownPrivate, peerPublic)
            ?? throw new CryptographicException(
                "X25519-Schluesselaustausch fehlgeschlagen - ungueltiger oeffentlicher Schluessel " +
                "(Punkt kleiner Ordnung?). Die Verbindung wird abgebrochen.");

        return shared.Export(SharedSecretBlobFormat.RawSharedSecret);
    }

    /// <summary>
    /// Fuenf Emojis aus einer 64er-Tabelle = 30 Bit. Beide Seiten zeigen dieselben
    /// Symbole; die Nutzer vergleichen sie ueber einen ANDEREN Kanal.
    /// Ein MitM muesste zwei verschiedene Sitzungsschluessel etablieren, die
    /// zufaellig dieselben fuenf Emojis ergeben: 1 zu 2^30.
    /// Siehe docs/05-security.md §5.4.
    /// </summary>
    public static (string Emoji, int[] Indices) ComputeSas(ReadOnlySpan<byte> sessionKey)
    {
        var raw = Hkdf(sessionKey, ReadOnlySpan<byte>.Empty, "ac/1 sas", 5);
        var indices = new int[5];
        var sb = new StringBuilder();
        for (var i = 0; i < 5; i++)
        {
            indices[i] = raw[i] % 64;
            if (i > 0) sb.Append(' ');
            sb.Append(EmojiTable.Symbols[indices[i]]);
        }
        return (sb.ToString(), indices);
    }

    /// <summary>Anzeigefreundlicher Fingerprint eines statischen Schluessels.</summary>
    public static string Fingerprint(ReadOnlySpan<byte> staticPublicKey)
    {
        var hash = SHA256.HashData(staticPublicKey);
        var sb = new StringBuilder();
        for (var i = 0; i < 10; i++)
        {
            if (i > 0 && i % 2 == 0) sb.Append(' ');
            sb.Append(hash[i].ToString("x2"));
        }
        return sb.ToString();
    }

    internal static byte[] Hkdf(ReadOnlySpan<byte> ikm, ReadOnlySpan<byte> salt, string info, int length)
    {
        // HKDF.DeriveKey hat zwei Ueberladungen: eine mit byte[] und
        // Rueckgabewert, eine mit Spans und Ausgabepuffer. Mischen geht nicht -
        // unsere Parameter sind Spans, also die Span-Variante.
        var output = new byte[length];
        HKDF.DeriveKey(HashAlgorithmName.SHA256, ikm, output, salt, Encoding.UTF8.GetBytes(info));
        return output;
    }

    public static Key GenerateEphemeral()
        => Key.Create(X25519, new KeyCreationParameters
        {
            ExportPolicy = KeyExportPolicies.AllowPlaintextExport,
        });

    public static byte[] PublicKeyBytes(Key key)
        => key.PublicKey.Export(KeyBlobFormat.RawPublicKey);

    public static readonly JsonSerializerOptions JsonOptions = new()
    {
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };
}
